homeassistant/components/ecovacs/entity.py

```diff
@@ -0,0 +1,113 @@
+"""Ecovacs mqtt entity module."""
+from __future__ import annotations
+
+from collections.abc import Callable, Coroutine
+from dataclasses import dataclass
+from typing import Any, Generic, TypeVar
+
+from deebot_client.capabilities import Capabilities
+from deebot_client.device import Device
+from deebot_client.events import AvailabilityEvent
+from deebot_client.events.base import Event
+
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity import Entity, EntityDescription
+
+from .const import DOMAIN
+
+_EntityDescriptionT = TypeVar("_EntityDescriptionT", bound=EntityDescription)
+CapabilityT = TypeVar("CapabilityT")
+EventT = TypeVar("EventT", bound=Event)
+
+
+@dataclass(kw_only=True, frozen=True)
+class EcovacsEntityDescription(
+    EntityDescription,
+    Generic[CapabilityT],
+):
+    """Ecovacs entity description."""
+
+    capability_fn: Callable[[Capabilities], CapabilityT | None]
+
+
+class EcovacsEntity(Entity, Generic[CapabilityT, _EntityDescriptionT]):
+    """Ecovacs entity."""
+
+    entity_description: _EntityDescriptionT
+
+    _attr_should_poll = False
+    _attr_has_entity_name = True
+    _always_available: bool = False
+
+    def __init__(
+        self,
+        device: Device,
+        capability: CapabilityT,
+        entity_description: _EntityDescriptionT | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize entity."""
+        super().__init__(**kwargs)
+        if entity_description:
+            self.entity_description = entity_description
+        elif not hasattr(self, "entity_description"):
+            raise ValueError(
+                '"entity_description" must be either set as class variable or passed on init!'
+            )
```

Comment:
Why do we actually do it this way?

Reply:
Some entities (which will be added in a follow-up) will pass the entity description as an init argument. Some other entities, like the vacuum, have set it as a class variable as it will not be dynamic.

I can change if you want to pass it always as init argument